compiler/noirc_evaluator/src/ssa/opt/independent_subgraph_detection.rs

```diff
@@ -0,0 +1,383 @@
+                        Value::Intrinsic(intrinsic) => match intrinsic {
+                            Intrinsic::IsUnconstrained
+                            | Intrinsic::AsWitness
+                            | Intrinsic::ApplyRangeConstraint
+                            | Intrinsic::AssertConstant => {}
+                            _ => {
+                                let mut value_ids = argument_ids.clone();
+                                value_ids.extend_from_slice(results);
+                                self.connect_values(function, &value_ids);
+                            }
+                        },
+                        Value::Function(callee) => match all_functions[&callee].runtime() {
+                            RuntimeType::Brillig => {
+                                // For calls to brillig functions we memorize the mapping of results to argument ValueId's and InstructionId's
+                                // The latter are needed to produce the callstack later
+                                for result in results.iter() {
+                                    self.brillig_return_to_argument_map
+                                        .insert(*result, argument_ids.clone());
+                                    self.brillig_return_to_instruction_id_map
+                                        .insert(*result, *instruction);
+                                }
+                            }
+                            _ => {
+                                let mut value_ids = argument_ids.clone();
+                                value_ids.extend_from_slice(results);
+                                self.connect_values(function, &value_ids);
+                            }
+                        },
+                        Value::ForeignFunction(..) => {
+                            panic!("Should not be able to reach foreign function from non-brillig functions");
+                        }
+                        _ => {
+                            panic!("At the point we are running disconnect there shouldn't be any other values as arguments")
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        self.block_queue.extend(function.dfg[block].successors());
+    }
+
+    /// Add a set of ValueIds to the vector of connected values while ignoring constants
+    fn connect_values(&mut self, function: &Function, values: &[ValueId]) {
+        self.value_sets.push(HashSet::from_iter(
+            values
+                .iter()
+                .filter(|value_id| {
+                    !matches!(function.dfg.get_value(**value_id), Value::NumericConstant { .. })
+                })
+                .cloned(),
+        ));
+    }
+
+    /// Merge all small sets into larger ones based on whether the sets intersect or not
+    ///
+    /// If two small sets have a common ValueId, we merge them into one
+    fn merge_sets(&mut self) {
```

Comment:
I think this would be simpler if when we inserted these to begin with they were already merged.
E.g. before inserting a new set of ids, check if any id is in an existing set then add them to that one.

Merging them after the fact like this seems to require a lot of upkeep of merging sets and removing old sets.

Reply:
Can I rely on ValueIds not appearing out of line here? As in a ValueId will always appear in an instruction that connects it to previous value Ids first? Because with brillig calls it seems that I can't and then I have to merge sets anyway

Reply:
Spoke about this on slack a bit: it's possible we can do this as we go but uncertain if it'll be shorter or not. It is not worth blocking over.